File: stockdownloader/backtest/options_backtest_result.py
"""Stores options backtest results and computes performance metrics.

Includes total P/L from premiums, win rate, average premium captured,
max drawdown, and volume statistics.
"""

import math
from stockdownloader.model.options_trade import OptionsTrade, OptionsTradeStatus
from stockdownloader.model.price_data import PriceData


class OptionsBacktestResult:
    def __init__(self, strategy_name: str, initial_capital: float):
        self.strategy_name = strategy_name
        self.initial_capital = initial_capital
        self.final_capital = initial_capital
        self._trades: list[OptionsTrade] = []
        self.equity_curve: list[float] = []
        self.start_date: str = ""
        self.end_date: str = ""
        self.total_volume_traded: int = 0

    def add_trade(self, trade: OptionsTrade) -> None:
        self._trades.append(trade)
        self.total_volume_traded += trade.entry_volume

    @property
    def closed_trades(self) -> list[OptionsTrade]:
        return [t for t in self._trades if t.status != OptionsTradeStatus.OPEN]

    @property
    def total_return(self) -> float:
        if self.initial_capital == 0:
            return 0.0
        return ((self.final_capital - self.initial_capital) / self.initial_capital) * 100

    @property
    def total_profit_loss(self) -> float:
        return self.final_capital - self.initial_capital

    @property
    def total_trades(self) -> int:
        return len(self.closed_trades)

    @property
    def winning_trades(self) -> int:
        return sum(1 for t in self.closed_trades if t.is_win)

    @property
    def losing_trades(self) -> int:
        return self.total_trades - self.winning_trades

    @property
    def win_rate(self) -> float:
        if self.total_trades == 0:
            return 0.0
        return (self.winning_trades / self.total_trades) * 100

    @property
    def average_win(self) -> float:
        wins = [t.profit_loss for t in self.closed_trades if t.is_win]
        return sum(wins) / len(wins) if wins else 0.0

    @property
    def average_loss(self) -> float:
        losses = [t.profit_loss for t in self.closed_trades if not t.is_win]
        return sum(losses) / len(losses) if losses else 0.0

    @property
    def profit_factor(self) -> float:
        gross_profit = sum(t.profit_loss for t in self.closed_trades if t.is_win)
        gross_loss = sum(abs(t.profit_loss) for t in self.closed_trades if not t.is_win)
        if gross_loss == 0:
            return 999.99 if gross_profit > 0 else 0.0
        return round(gross_profit / gross_loss, 2)

    @property
    def max_drawdown(self) -> float:
        if not self.equity_curve:
            return 0.0
        peak = self.equity_curve[0]
        max_dd = 0.0
        for equity in self.equity_curve:
            if equity > peak:
                peak = equity
            dd = ((peak - equity) / peak) * 100 if peak > 0 else 0.0
            if dd > max_dd:
                max_dd = dd
        return max_dd

    def sharpe_ratio(self, trading_days_per_year: int = 252) -> float:
        if len(self.equity_curve) < 2:
            return 0.0
        daily_returns = []
        for i in range(1, len(self.equity_curve)):
            prev = self.equity_curve[i - 1]
            if prev != 0:
                daily_returns.append((self.equity_curve[i] - prev) / prev)
        if not daily_returns:
            return 0.0
        mean_return = sum(daily_returns) / len(daily_returns)
        sum_sq = sum((r - mean_return) ** 2 for r in daily_returns)
        std_dev = math.sqrt(sum_sq / len(daily_returns))
        if std_dev == 0:
            return 0.0
        return round((mean_return / std_dev) * math.sqrt(trading_days_per_year), 2)

    def buy_and_hold_return(self, data: list[PriceData]) -> float:
        if not data:
            return 0.0
        first = data[0].close
        last = data[-1].close
        return ((last - first) / first) * 100 if first != 0 else 0.0

    @property
    def average_premium_collected(self) -> float:
        """Average premium collected per trade (for short strategies)."""
        closed = self.closed_trades
        if not closed:
            return 0.0
        total_premium = sum(t.total_entry_cost() for t in closed)
        return round(total_premium / len(closed), 2)

    @property
    def trades(self) -> list[OptionsTrade]:
        return list(self._trades)

File: src/main/java/com/stockdownloader/backtest/OptionsBacktestResult.java
package com.stockdownloader.backtest;

import com.stockdownloader.model.OptionsTrade;
import com.stockdownloader.model.PriceData;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/**
 * Stores options backtest results and computes performance metrics including
 * total P/L from premiums, win rate, average premium captured, max drawdown,
 * and volume statistics.
 */
public final class OptionsBacktestResult {

    private static final BigDecimal HUNDRED = BigDecimal.valueOf(100);

    private final String strategyName;
    private final BigDecimal initialCapital;
    private final List<OptionsTrade> trades = new ArrayList<>();
    private BigDecimal finalCapital;
    private List<BigDecimal> equityCurve = List.of();
    private String startDate;
    private String endDate;
    private long totalVolumeTraded;

    public OptionsBacktestResult(String strategyName, BigDecimal initialCapital) {
        this.strategyName = Objects.requireNonNull(strategyName);
        this.initialCapital = Objects.requireNonNull(initialCapital);
        this.finalCapital = initialCapital;
    }

    public void addTrade(OptionsTrade trade) {
        trades.add(Objects.requireNonNull(trade));
        totalVolumeTraded += trade.getEntryVolume();
    }

    public void setFinalCapital(BigDecimal v) { this.finalCapital = v; }
    public void setEquityCurve(List<BigDecimal> curve) { this.equityCurve = List.copyOf(curve); }
    public void setStartDate(String d) { this.startDate = d; }
    public void setEndDate(String d) { this.endDate = d; }

    public BigDecimal getTotalReturn() {
        return finalCapital.subtract(initialCapital)
                .divide(initialCapital, 6, RoundingMode.HALF_UP)
                .multiply(HUNDRED);
    }

    public BigDecimal getTotalProfitLoss() {
        return finalCapital.subtract(initialCapital);
    }

    public List<OptionsTrade> getClosedTrades() {
        return trades.stream()
                .filter(t -> t.getStatus() != OptionsTrade.Status.OPEN)
                .toList();
    }

    public long getTotalTrades() { return getClosedTrades().size(); }

    public long getWinningTrades() {
        return getClosedTrades().stream().filter(OptionsTrade::isWin).count();
    }

    public long getLosingTrades() { return getTotalTrades() - getWinningTrades(); }

    public BigDecimal getWinRate() {
        long total = getTotalTrades();
        if (total == 0) return BigDecimal.ZERO;
        return BigDecimal.valueOf(getWinningTrades())
                .divide(BigDecimal.valueOf(total), 4, RoundingMode.HALF_UP)
                .multiply(HUNDRED);
    }

    public BigDecimal getAverageWin() {
        return averagePL(true);
    }

    public BigDecimal getAverageLoss() {
        return averagePL(false);
    }

    private BigDecimal averagePL(boolean winners) {
        var filtered = getClosedTrades().stream()
                .filter(t -> t.isWin() == winners)
                .map(OptionsTrade::getProfitLoss)
                .toList();
        if (filtered.isEmpty()) return BigDecimal.ZERO;
        BigDecimal sum = filtered.stream().reduce(BigDecimal.ZERO, BigDecimal::add);
        return sum.divide(BigDecimal.valueOf(filtered.size()), 2, RoundingMode.HALF_UP);
    }

    public BigDecimal getProfitFactor() {
        BigDecimal grossProfit = BigDecimal.ZERO;
        BigDecimal grossLoss = BigDecimal.ZERO;
        for (OptionsTrade t : getClosedTrades()) {
            if (t.isWin()) {
                grossProfit = grossProfit.add(t.getProfitLoss());
            } else {
                grossLoss = grossLoss.add(t.getProfitLoss().abs());
            }
        }
        if (grossLoss.compareTo(BigDecimal.ZERO) == 0) {
            return grossProfit.compareTo(BigDecimal.ZERO) > 0
                    ? BigDecimal.valueOf(999.99) : BigDecimal.ZERO;
        }
        return grossProfit.divide(grossLoss, 2, RoundingMode.HALF_UP);
    }

    public BigDecimal getMaxDrawdown() {
        if (equityCurve.isEmpty()) return BigDecimal.ZERO;
        BigDecimal peak = equityCurve.getFirst();
        BigDecimal maxDD = BigDecimal.ZERO;
        for (BigDecimal equity : equityCurve) {
            if (equity.compareTo(peak) > 0) peak = equity;
            BigDecimal dd = peak.subtract(equity)
                    .divide(peak, 6, RoundingMode.HALF_UP)
                    .multiply(HUNDRED);
            if (dd.compareTo(maxDD) > 0) maxDD = dd;
        }
        return maxDD;
    }

    public BigDecimal getSharpeRatio(int tradingDaysPerYear) {
        if (equityCurve.size() < 2) return BigDecimal.ZERO;
        double[] dailyReturns = new double[equityCurve.size() - 1];
        for (int i = 1; i < equityCurve.size(); i++) {
            dailyReturns[i - 1] = equityCurve.get(i).subtract(equityCurve.get(i - 1))
                    .divide(equityCurve.get(i - 1), 10, RoundingMode.HALF_UP)
                    .doubleValue();
        }
        double mean = 0;
        for (double r : dailyReturns) mean += r;
        mean /= dailyReturns.length;
        double sumSqDiff = 0;
        for (double r : dailyReturns) sumSqDiff += Math.pow(r - mean, 2);
        double stdDev = Math.sqrt(sumSqDiff / dailyReturns.length);
        if (stdDev == 0) return BigDecimal.ZERO;
        double sharpe = (mean / stdDev) * Math.sqrt(tradingDaysPerYear);
        return BigDecimal.valueOf(sharpe).setScale(2, RoundingMode.HALF_UP);
    }

    /**
     * Average premium collected per trade (for short strategies).
     */
    public BigDecimal getAveragePremiumCollected() {
        var closed = getClosedTrades();
        if (closed.isEmpty()) return BigDecimal.ZERO;
        BigDecimal totalPremium = closed.stream()
                .map(OptionsTrade::totalEntryCost)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
        return totalPremium.divide(BigDecimal.valueOf(closed.size()), 2, RoundingMode.HALF_UP);
    }

    public String getStrategyName() { return strategyName; }
    public BigDecimal getInitialCapital() { return initialCapital; }
    public BigDecimal getFinalCapital() { return finalCapital; }
    public List<OptionsTrade> getTrades() { return Collections.unmodifiableList(trades); }
    public List<BigDecimal> getEquityCurve() { return equityCurve; }
    public String getStartDate() { return startDate; }
    public String getEndDate() { return endDate; }
    public long getTotalVolumeTraded() { return totalVolumeTraded; }
}
